### ISSUE TITLE
CSRF fails if mbstring.encoding_translation set to On

### ISSUE THREAD

**Issue by si3u** [yiisoft/yii2#14302]
### What steps will reproduce the problem?
I config php.ini with information bellow:
```ini
[mbstring]
mbstring.language = Japanese
mbstring.internal_encoding = UTF-8
mbstring.encoding_translation = On
```
### What is the expected result?
When I submit a form on website, I seen error

```
Error Bad Request (#400) Unable to verify your data submission
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.12
| PHP version      |  7.0.20
| Operating system | Windows

**Comment by cebe** [comment]
looks like an error with CSRF instead of PHP encoding configuration. what exactly is the problem?

**Comment by Ghiya** [comment]
I have the similar problem after updating to version 2.0.12.
In my case, as i find out by log traces, the issue is in CSRF cookie value validation.
Although the cookie is presented in `$_COOKIES` array as it has to be, [its value fails validation](https://github.com/yiisoft/yii2/blob/8ae207c3a186d7b4845a0e022d784ce0002dba8c/framework/web/Request.php#L1285) in `loadCookies` method of `yii\web\Request` class.
And in that case, the required cookie wont be presented in `$_cookies` array property and [new token will be generated](https://github.com/yiisoft/yii2/blob/8ae207c3a186d7b4845a0e022d784ce0002dba8c/framework/web/Request.php#L1322) in `getCsrfToken` method [according to condition](https://github.com/yiisoft/yii2/blob/8ae207c3a186d7b4845a0e022d784ce0002dba8c/framework/web/Request.php#L1325), [in its turn](https://github.com/yiisoft/yii2/blob/8ae207c3a186d7b4845a0e022d784ce0002dba8c/framework/web/Request.php#L1409) method `validateCsrfToken` will get the different `$trueToken` value and  [validation will fail](https://github.com/yiisoft/yii2/blob/8ae207c3a186d7b4845a0e022d784ce0002dba8c/framework/web/Request.php#L1414).

**Comment by Ghiya** [comment]
More specificly, hashed CSRF cookie data value `$calculatedHash` [differs form its stored hash value](https://github.com/yiisoft/yii2/blob/master/framework/base/Security.php#L429) `$hash`, so validation fails in `yii\base\Security`.
As i checked, in my case it happens in updated version, in the previous version CSRF cookie data value differs from the updated one and data validation works as it should.

In log traces it looks like:

```
yii\base\Security::validateData

Validating CSRF cookie : d4ae7ff74b0168747e431594a507418878be92ecdb3d8b1514c4bb92556085e9a:2:{i:0;s:13:"_csrf-backend";i:1;s:32:"F??R???֌?q???g8??`?????|yn?";}
```

```
yii\base\Security::validateData

Cookie `$hash` value:
d4ae7ff74b0168747e431594a507418878be92ecdb3d8b1514c4bb92556085e9
```

```
yii\base\Security::validateData

Cookie `$pureData` value:
a:2:{i:0;s:13:"_csrf-backend";i:1;s:32:"F??R???֌?q???g8??`?????|yn?";}
```

```
yii\base\Security::validateData

CSRF cookie `$hash` value differs from calculated `hash_hmac($pureData)` value
d4ae7ff74b0168747e431594a507418878be92ecdb3d8b1514c4bb92556085e9 : 91ff93ed84df7a2a56067cf085ac3c6d53aba1f60ad99914d6dd85ea0df860c4
returns : false
```

```
yii\web\Request::loadCookies

The value for the `_csrf-backend` is not validated.
```

**Comment by samdark** [comment]
@Ghiya do you have AJAX/redirects involved?

**Comment by si3u** [comment]
@samdark I use fresh source install via composer. Then, I submit to Register, Login form.

**Comment by samdark** [comment]
By fresh source you bean basic project template?

**Comment by si3u** [comment]
I use advanced project template.

**Comment by Ghiya** [comment]
@samdark It happens in any kind of request with CSRF validation and I have no AJAX or redirects involved.
For example, POST request from user authenthication form.
I'm using advanced project template too.
Seems like the issue is in cookie value control hash, cause data validation fails at the moment of its checking. However, my knowledge of the framework is not enough to draw certain conclusions.

**Comment by samdark** [comment]
Could you try basic template in the same environment to be sure it's not configuration specific?

**Comment by si3u** [comment]
@samdark Because I config environment in the php.ini file with information bellow

`
[mbstring]
mbstring.language = Japanese
mbstring.internal_encoding = UTF-8
mbstring.encoding_translation = On
`

Can you try it? I sure you will have to see this problem

**Comment by si3u** [comment]
I confirmed on basic template in the same environment, this problem appear.

**Comment by si3u** [comment]
$_COOKIE

| Name                | Value
| ---------------- | ---
|_csrf            |'83bc7498c79b4024059ab35ce118e844cbc5dc0c7b8bce92a705c1b67a216937a:2:{i:0;s:5:\"_csrf\";i:1;s:32:\"Q?*?Auj??h?(碻\0o? 9V????C\";}'

**Comment by samdark** [comment]
@Ghiya do you change mbstring settings?

**Comment by si3u** [comment]
@samdark I just try comment two lines

```
 [mbstring]
 //mbstring.language = Japanese
 //mbstring.internal_encoding = UTF-8
 mbstring.encoding_translation = On
```

I think because this config

```
mbstring.encoding_translation = On
```

**Comment by si3u** [comment]
![image](https://user-images.githubusercontent.com/28697263/27493193-13620706-5873-11e7-99b1-b9f464b65a7a.png)

![image](https://user-images.githubusercontent.com/28697263/27493239-3cb5dd12-5873-11e7-9c56-b15909862df2.png)

**Comment by Ghiya** [comment]
@samdark In my case, setting `mbstring.encoding_translation` to the `off` value throws exception `InvalidParamException` for json encoding, with standard message `Malformed UTF-8 characters, possibly incorrectly encoded.`

Here is actual `mbstring` settings:
![2017-06-23 22 08 48](https://user-images.githubusercontent.com/8424672/27496892-98cb4032-5860-11e7-9ed0-9e1bb0c18bef.png)

**Comment by samdark** [comment]
@Ghiya your case is different. Try turning off `func_overload`.

**Comment by samdark** [comment]
@si3u what's on the screenshot is OK. Value in the cookie is masked w/ HMAC.

**Comment by cebe** [comment]
I had one case where people were complaining about not being able to log cookie values because db target does not allow invalid UTF8 to be stored in pgsql table. maybe we should base64encode the value to avoid invalid chars?

**Comment by samdark** [comment]
Wouldn't hurt, I think.

**Comment by samdark** [comment]
`mbstring.encoding_translation = on` causes it.

**Comment by cebe** [comment]
should be fixed by #14590
